### DISCUSSION TITLE
Assigning different material properties to gauss point

### DISCUSSION THREAD

**Discussion by manikumd22** [idaholab/moose#28562] [Q&A Modules: Solid mechanics]
Hi @GiudGiud 

I am working on a project that requires varying the material properties for different gauss points.

For instance 2D plate will have Gauss points, and I want to assign different materials to each Gauss point in order to increase x 

For instance,
```
[Materials]
[compute_stress1]
type = ComputeNeoHookeanStress
lambda = 40.0
mu = 67.0
x = 0.1, all y
[]
[compute_stress2]
type = ComputeNeoHookeanStress
lambda = 400.0
mu = 670.0
x =0.2, all y
[]
[compute_stress3]
type = ComputeNeoHookeanStress
lambda = 4000.0
mu = 6700.0
x = 0.3, all y
[]
[compute_strain]
type = ComputeLagrangianStrain
[]
[]
```
and so on
![x_vary](https://github.com/user-attachments/assets/5300f5da-9eea-4ae2-b26f-ee2e6e346822)

I am struggling with that as I don't know the location of gauss points coordinate

Or
```
  **if (_large_kinematics)
  {
      if ( _q_point[_qp](1) = 0.1;) {
        RankTwoTensor Cinv = (2 * _E[_qp] + RankTwoTensor::Identity()).inverse();
        _S[_qp] = 100* (_lambda[_qp] * log(_F[_qp].det()) - _mu[_qp]) * Cinv +
              _mu[_qp] * RankTwoTensor::Identity();
        _C[_qp] =100*(
            -2 * (_lambda[_qp] * log(_F[_qp].det()) - _mu[_qp]) * Cinv.times<i_, k_, l_, j_>(Cinv) +
            _lambda[_qp] * Cinv.times<i_, j_, k_, l_>(Cinv));
    }  
      if ( _q_point[_qp](1) = 0.2;) {
        RankTwoTensor Cinv = (2 * _E[_qp] + RankTwoTensor::Identity()).inverse();
        _S[_qp] = 1000* (_lambda[_qp] * log(_F[_qp].det()) - _mu[_qp]) * Cinv +
              _mu[_qp] * RankTwoTensor::Identity();
        _C[_qp] =1000*(
            -2 * (_lambda[_qp] * log(_F[_qp].det()) - _mu[_qp]) * Cinv.times<i_, k_, l_, j_>(Cinv) +
            _lambda[_qp] * Cinv.times<i_, j_, k_, l_>(Cinv));
    }  
      if ( _q_point[_qp](1) = 0.3;){
        RankTwoTensor Cinv = (2 * _E[_qp] + RankTwoTensor::Identity()).inverse();
        _S[_qp] = 10000* (_lambda[_qp] * log(_F[_qp].det()) - _mu[_qp]) * Cinv +
              _mu[_qp] * RankTwoTensor::Identity();
        _C[_qp] =10000*(
            -2 * (_lambda[_qp] * log(_F[_qp].det()) - _mu[_qp]) * Cinv.times<i_, k_, l_, j_>(Cinv) +
            _lambda[_qp] * Cinv.times<i_, j_, k_, l_>(Cinv));
    }  
  }**
```

**Comment by GiudGiud** [comment]
Hello

We dont have very objects to output the quadrature points. 
I think the cleanest might be to create a QuadraturePointsPositions object and get the JSON output from there.
I ll get started on this right now

**Reply by GiudGiud** [reply]
https://github.com/idaholab/moose/pull/28565 has a new Positions object that will output the quadrature points to JSON

**Reply by manikumd22** [reply]
Hi @GiudGiud 
I tried as per that discussion, but I am getting an error

*** ERROR ***
A 'QuadraturePointsPositions' is not a registered object.

If you are trying to find this object in a dynamically loaded library, make sure that
the library can be found either in your "Problem/library_path" parameter or in the
MOOSE_LIBRARY_PATH environment variable.[39m





#Tensor Mechanics tutorial: the basics
#Step 1, part 1
#2D simulation of uniaxial tension with linear elasticity

[GlobalParams]
    displacements = 'disp_x disp_y'
[]
  
[Mesh]
  [msh]
    type = GeneratedMeshGenerator
    dim = 2
    nx = 4
    ny = 4
  []
[]

[Positions]
  [all_mesh]
    type = QuadraturePointsPositions
    auto_sort = true
  []
[]
  
[Modules/TensorMechanics/Master]
[all]
add_variables = true
strain = small
generate_output = 'vonmises_stress stress_xx stress_yy strain_xx strain_yy'
[]
[]
  
[Materials]

[elasticity1]
    type = ComputeIsotropicElasticityTensor
    youngs_modulus = 9e4
    poissons_ratio = 0.3
[]

[./stress]
    type = ComputeLinearElasticStress
[../]
[]
  
[BCs]
[./left]
    type = DirichletBC
    variable = disp_x
    boundary = 'top bottom'
    value = 0.0
[../]
[./bottom]
    type = DirichletBC
    variable = disp_y
    boundary = 'top'
    value = 1
[../]
[./top]
    type = DirichletBC
    variable = disp_y
    boundary = 'bottom'
    value = 0
[../]
[]
  
[Preconditioning]
[./SMP]
    type = SMP
    full = true
[../]
[]

[Postprocessors]
  [num_pos]
    type = NumPositions
    positions = all_mesh
  []
[]

[Executioner]
type = Steady

solve_type = 'NEWTON'

petsc_options = '-snes_ksp_ew'
petsc_options_iname = '-pc_type -sub_pc_type -pc_asm_overlap -ksp_gmres_restart'
petsc_options_value = 'asm lu 1 101'
[]
  
[Outputs]
type = JSON
exodus = true
perf_graph = true
file_base = ./inputfiles/maniklinearelastic_ir_out12
[]

**Reply by GiudGiud** [reply]
did you pull that branch and built it?
You can either do that or add the source and header to your local installation then rebuild

**Reply by GiudGiud** [reply]
for clarification, this object is not in the moose source code yet, I just made it this afternoon. So you have to either use my branch or add it to your code

**Comment by manikumd22** [comment]
Hi @GiudGiud 

I add it to my local moose app, 

![Screenshot 2024-09-06 at 5 29 08 PM (2)](https://github.com/user-attachments/assets/b69726a9-d6e4-48d8-92e5-1657463054d4)

**Reply by GiudGiud** [reply]
did you recompile?

**Reply by manikumd22** [reply]
yes

**Reply by GiudGiud** [reply]
umm.
then I guess:
- save all your work, include this new object, with git commit
- go in the top folder of your app
- git clean -Xfd
- recompile with make -j 6
- try again

**Reply by manikumd22** [reply]
Hi @GiudGiud 

I tried but getting the same error

**Reply by GiudGiud** [reply]
you can try to re-register the object by changing the registerMooseObject call in QuadraturePointsPositions.C

when you do, check that the object is being properly recompiled

**Reply by manikumd22** [reply]
Hi @GiudGiud ,


Now it worked,

thank you so much


For example, I'm getting in this form

                    {
                        "x": 0.052831216351296784,
                        "y": 0.052831216351296784,
                        "z": 0.0
                    },
                    {
                        "x": 0.052831216351296784,
                        "y": 0.19716878364870322,
                        "z": 0.0
                    },
                    {
                        "x": 0.19716878364870322,
                        "y": 0.9471687836487033,
                        "z": 0.0
                    },
                    {
                        "x": 0.30283121635129684,
                        "y": 0.052831216351296784,
                        "z": 0.0
                    },


Will this below material module work, if I define it in this way and will  _q_point[_qp](1) =  0.052831216351296784 take all the 9 gauss points corresponding to this or I have to define both _q_point[_qp](1) =  0.052831216351296784 and _q_point[_qp](2) =  0.052831216351296784

  **if (_large_kinematics)
  {
      if ( _q_point[_qp](1) = 0.052831216351296784;) {
        RankTwoTensor Cinv = (2 * _E[_qp] + RankTwoTensor::Identity()).inverse();
        _S[_qp] = 100* (_lambda[_qp] * log(_F[_qp].det()) - _mu[_qp]) * Cinv +
              _mu[_qp] * RankTwoTensor::Identity();
        _C[_qp] =100*(
            -2 * (_lambda[_qp] * log(_F[_qp].det()) - _mu[_qp]) * Cinv.times<i_, k_, l_, j_>(Cinv) +
            _lambda[_qp] * Cinv.times<i_, j_, k_, l_>(Cinv));
    }  
      if ( _q_point[_qp](1) = 0.19716878364870322;) {
        RankTwoTensor Cinv = (2 * _E[_qp] + RankTwoTensor::Identity()).inverse();
        _S[_qp] = 1000* (_lambda[_qp] * log(_F[_qp].det()) - _mu[_qp]) * Cinv +
              _mu[_qp] * RankTwoTensor::Identity();
        _C[_qp] =1000*(
            -2 * (_lambda[_qp] * log(_F[_qp].det()) - _mu[_qp]) * Cinv.times<i_, k_, l_, j_>(Cinv) +
            _lambda[_qp] * Cinv.times<i_, j_, k_, l_>(Cinv));
    }  
      if ( _q_point[_qp](1) = 0.30283121635129684;){
        RankTwoTensor Cinv = (2 * _E[_qp] + RankTwoTensor::Identity()).inverse();
        _S[_qp] = 10000* (_lambda[_qp] * log(_F[_qp].det()) - _mu[_qp]) * Cinv +
              _mu[_qp] * RankTwoTensor::Identity();
        _C[_qp] =10000*(
            -2 * (_lambda[_qp] * log(_F[_qp].det()) - _mu[_qp]) * Cinv.times<i_, k_, l_, j_>(Cinv) +
            _lambda[_qp] * Cinv.times<i_, j_, k_, l_>(Cinv));
    }  
  }**

**Reply by GiudGiud** [reply]
no this wont work
```
if ( _q_point[_qp](https://github.com/idaholab/moose/discussions/1) = 0.052831216351296784;) {
```
this sets (single = instead of two ==) the value of the q point
AND
the quadrature point will never be exactly equal to ` 0.052831216351296784`, you need a tolerance.

So instead
```
if (MooseUtils::absoluteFuzzyEqual(_q_point[_qp](1), 0.052831216351296784))
```
this used a 1e-12 tolerance iirc, enough here
